### PR TITLE
Exceptions in dbe - to catch properly #1884

### PR DESCRIPTION
**Pull request by igalink**
Propogate c++ Exceptions to Python

### CODE REVIEW

File: Embedded/Python/dbe.pyx [hunk 1]
```diff
@@ -159,85 +159,146 @@ ColumnDetailsTp = namedtuple("ColumnDetails", ["name", "type", "nullable",
...
-            assert not self.closed
+            if self.closed:
+                raise RuntimeError('Initialization failed')
         except OSError as err:
-            print("DBEngine: OS error: {0}".format(err))
-            raise
-        except ValueError:
-            print("DBEngine: ValueError: Could not convert data to string")
-            raise
+            print("DBE init: OS error: {0}".format(err))
```
Comment (ienkovich) [inline Embedded/Python/dbe.pyx:173]: Please remove all these prints. We need to forward exceptions to the user code and shouldn't print anything.

**Review by ienkovich — CHANGES_REQUESTED** [review]

File: Embedded/Python/dbe.pyx (hunk 1 above)
Comment (igalink) [reply Embedded/Python/dbe.pyx]: Done

File: Embedded/DBEngine.cpp [hunk 2]
```diff
@@ -220,80 +183,63 @@ class DBEngineImpl : public DBEngine {
...
+          // skip shards, they're not standalone tables
+          continue;
         }
-      } catch (std::exception const& e) {
-        std::cerr << "DBE:getTables: " << e.what() << std::endl;
+        table_names.push_back(td->tableName);
       }
     } else {
-      std::cerr << "DBE:getTables: catalog is NULL" << std::endl;
+      throw std::runtime_error("System catalog uninitialized");
```
Comment (ienkovich) [inline Embedded/DBEngine.cpp:242]: Can we really get here? I don't see how successfully initialized engine may have no catalog.

File: Embedded/DBEngine.cpp [hunk 3]
```diff
@@ -550,22 +504,17 @@ inline const DBEngineImpl* getImpl(const DBEngine* ptr) {
...
-}
-
 void DBEngine::executeDDL(const std::string& query) {
   DBEngineImpl* engine = getImpl(this);
   engine->executeDDL(query);
 }
 
-std::unique_ptr<Cursor> DBEngine::executeDML(const std::string& query) {
+std::shared_ptr<Cursor> DBEngine::executeDML(const std::string& query) {
   DBEngineImpl* engine = getImpl(this);
```
Comment (ienkovich) [inline Embedded/DBEngine.cpp:513]: Not related to this PR but why do we need this cast?

File: Embedded/Python/dbe.pyx [hunk 4]
```diff
@@ -191,47 +176,50 @@ cdef class PyDbEngine:
...
-        obj.c_cursor = self.c_dbe.executeDML(bytes(query, 'utf-8'));
+        obj.c_cursor = self.c_dbe.get().executeDML(bytes(query, 'utf-8'));
         return obj;
 
     def executeRA(self, query):
-        obj = PyCursor();
-        assert not self.closed
-        obj.c_cursor = self.c_dbe.executeRA(bytes(query, 'utf-8'));
-        return obj
+            self.check_closed()
```
Comment (ienkovich) [inline Embedded/Python/dbe.pyx:194]: There is an extra tab

**Review by ienkovich — COMMENTED** [review]
Overall the patch looks good but it misses tests.

File: Embedded/DBEngine.cpp (hunk 2 above)
Comment (igalink) [reply Embedded/DBEngine.cpp:242]: this is unlikely, but it's better to check than to have a sigfault in python

File: Embedded/DBEngine.cpp (hunk 3 above)
Comment (igalink) [reply Embedded/DBEngine.cpp:513]: "except+" linkage in the DBEngine.pxd requares copy constructor

File: Embedded/Python/dbe.pyx (hunk 4 above)
Comment (igalink) [reply Embedded/Python/dbe.pyx]: Removed

File: Embedded/DBEngine.cpp (hunk 3 above)
Comment (ienkovich) [reply Embedded/DBEngine.cpp:513]: I meant `DBEngineImpl* engine = getImpl(this);`

**Review by ienkovich — CHANGES_REQUESTED** [review]
The test should be automated and integrated into DBE regression testing.

File: Embedded/DBEngine.cpp (hunk 3 above)
Comment (igalink) [reply Embedded/DBEngine.cpp:513]: to avoid virtual functions

Comment (ienkovich) [reply Embedded/DBEngine.cpp:513]: You have a weird structure where implementation inherits from the base class which is not actually an interface (why to inherit then?). This code will break if anyone will try to use `DBEngine` as a base class.

It looks to me you try to apply [pimpl](https://en.cppreference.com/w/cpp/language/pimpl) approach but in a weird way. Again, this is not for this PR but this should be fixed eventually (avoid static casts in the base class).

**Comment by igalink** [comment]
> The test should be automated and integrated into DBE regression testing.

test_exceptions.py changed to use PyTest

File: Embedded/test/test_exceptions.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+def test_failed_init():
+    try:
+        shutil.rmtree(data_path)
+    except FileNotFoundError:
+        pass
+
+    with pytest.raises(RuntimeError) as excinfo:
+        def f():
+            engine = dbe.PyDbEngine(**{"port": 5555, "path": "/"+data_path})
+            with pytest.raises(RuntimeError) as excinfo:
```
Comment (ienkovich) [inline Embedded/test/test_exceptions.py:24]: IIUC engine construction here raises an error. In this case, the following code will not be executed.

File: Embedded/test/test_exceptions.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+######################Check DDL statement
+def test_success_DDL():
+    engine.executeDDL("create table test (x int not null, w tinyint, y int, z text)")
+    assert engine.get_tables() == ['test']
+
+#######################Check creating a duplicate table
+def test_failed_DDL():
+    with pytest.raises(RuntimeError) as excinfo:
+        def f():
```
Comment (ienkovich) [inline Embedded/test/test_exceptions.py:48]: Why do you put 'with' blocks into these nested functions?

Comment (igalink) [reply Embedded/test/test_exceptions.py]: Removed

File: Embedded/test/test_exceptions.py (hunk 5 above)
Comment (igalink) [reply Embedded/test/test_exceptions.py]: Fixed

**Review by ienkovich — APPROVED** [review]

**Pull request merged**